crates/bevy_reflect/src/type_registry.rs

```diff
@@ -182,27 +182,30 @@ impl TypeRegistration {
     /// data.
     ///
     /// Returns `None` if no such value exists.
+    ///
+    /// The value is guaranteed to be a value of `T` generated using [`FromType`] of the type that this
+    /// registration is used for.
     pub fn data<T: TypeData>(&self) -> Option<&T> {
         self.data
             .get(&TypeId::of::<T>())
             .and_then(|value| value.downcast_ref())
     }
 
-    /// Returns a mutable reference to the value of type `T` in this
-    /// registration's type data.
+    /// Inserts an instance of `D` into this registration's type data.
+    /// The value is created using the [`FromType<T>`] impl for `D`
+    /// and `T` must match the type that this [`TypeRegistration`] was created for.
     ///
-    /// Returns `None` if no such value exists.
-    pub fn data_mut<T: TypeData>(&mut self) -> Option<&mut T> {
-        self.data
-            .get_mut(&TypeId::of::<T>())
-            .and_then(|value| value.downcast_mut())
-    }
-
-    /// Inserts an instance of `T` into this registration's type data.
-    ///
-    /// If another instance of `T` was previously inserted, it is replaced.
-    pub fn insert<T: TypeData>(&mut self, data: T) {
-        self.data.insert(TypeId::of::<T>(), Box::new(data));
+    /// If another instance of `D` was previously inserted, it is replaced.
+    pub fn insert<T: 'static, D: TypeData + FromType<T>>(&mut self) {
+        if self.type_id != std::any::TypeId::of::<T>() {
+            panic!(
+            "called `TypeRegistration::insert` on a registration for `{}` with data for type `{}`",
+            self.short_name,
```

Comment:
Maybe we use the full name here for clarity? There may be odd instances where a user is trying to insert for type `my_crate::Foo` but is accidentally using `lib_crate::Foo` (which would result in `called 'TypeRegistration::insert' on a registration for 'Foo' with data for type 'lib_crate::Foo'`.

Admittedly, this isn't that bad already, but it could be a bit clearer imo.

(Would need to update the panicking test as well.)